Editors/vscode/src/inlayHints.ts

```diff
@@ -0,0 +1,202 @@
+        const newSourceFiles = new Map<string, SourceFile>();
+
+        // Rerender all, even up-to-date editors for simplicity
+        this.visibleSourceKitLSPEditors.forEach(async editor => {
+            const uri = editor.document.uri.toString();
+            const file = this.sourceFiles.get(uri) ?? {
+                document: editor.document,
+                inFlightInlayHints: null,
+                cachedDecorations: null
+            };
+            newSourceFiles.set(uri, file);
+
+            // No text documents changed, so we may try to use the cache
+            if (!file.cachedDecorations) {
+                const hints = await this.fetchHints(file);
+                if (!hints) return;
+                file.cachedDecorations = this.hintsToDecorations(hints);
+            }
+
+            this.renderDecorations(editor, file.cachedDecorations);
+        });
+
+        // Cancel requests for no longer visible (disposed) source files
+        this.sourceFiles.forEach((file, uri) => {
+            if (!newSourceFiles.has(uri)) {
+                file.inFlightInlayHints?.cancel();
+            }
+        });
+
+        this.sourceFiles = newSourceFiles;
+    }
+
+    private onDidChangeTextDocument(event: vscode.TextDocumentChangeEvent): void {
+        if (event.contentChanges.length !== 0 && this.isSourceKitLSPDocument(event.document)) {
+            this.syncCacheAndRenderHints();
+        }
+    }
+
+    private syncCacheAndRenderHints(): void {
+        this.sourceFiles.forEach(async (file, uri) => {
+            const hints = await this.fetchHints(file);
+            if (!hints) return;
+
+            const decorations = this.hintsToDecorations(hints);
+            file.cachedDecorations = decorations;
+
+            this.visibleSourceKitLSPEditors.forEach(editor => {
+                if (editor.document.uri.toString() === uri) {
+                    this.renderDecorations(editor, decorations);
+                }
+            });
+        });
+    }
+
+    private get visibleSourceKitLSPEditors(): vscode.TextEditor[] {
+        return vscode.window.visibleTextEditors.filter(e => this.isSourceKitLSPDocument(e.document));
+    }
+
+    private isSourceKitLSPDocument(document: vscode.TextDocument): boolean {
+        // TODO: Add other SourceKit-LSP languages if/once we forward inlay
+        // hint requests to clangd.
+        return document.languageId === 'swift' && document.uri.scheme === 'file';
+    }
+
+    private renderDecorations(editor: vscode.TextEditor, decorations: vscode.DecorationOptions[]): void {
+        editor.setDecorations(hintStyle.decorationType, decorations);
+    }
+
+    private hintsToDecorations(hints: InlayHint[]): vscode.DecorationOptions[] {
+        const converter = this.client.protocol2CodeConverter;
+        return hints.map(h => hintStyle.makeDecoration(h, converter));
+    }
+
+    private async fetchHints(file: SourceFile): Promise<InlayHint[]> {
+        file.inFlightInlayHints?.cancel();
```

Comment:
This means, we can‘t currently fetch inlay hints for two files concurrently, right? Wouldn’t it be better to make `inFlightInlayHints` be a map `URI -> cancellation token` and only cancel the request for the file that we are now requesting new hints for?

It’s not really relevant yet since SourceKit supports neither cancellation nor concurrent requests, but it would be nice to be on the safe side now already.